tests/lib/lte_lc_api/src/lte_lc_api_test.c

```diff
@@ -0,0 +1,1261 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <unity.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <modem/lte_lc.h>
+#include <nrf_errno.h>
+#include "lte_lc_helpers.h"
+#include <mock_functions_to_mock.h>
+
+enum nrf_modem_at_scanf_variant_t {
+	INVALID = 0,
+	CEREG_lte_lc_nw_reg_status_get,
+	CEREG_lte_lc_lte_mode_get,
+	XSYSTEMMODE_lte_lc_system_mode_get,
+	CFUN_lte_lc_func_mode_get,
+	CONEVAL_lte_lc_conn_eval_params_get,
+	PERIODICSEARCHCONF_lte_lc_periodic_search_get,
+	AT_SCANF_FAIL,
+};
+
+struct nrf_modem_at_scanf_return_t {
+	int ltem_mode;
+	int nbiot_mode;
+	int gps_mode;
+	int mode_preference;
+	uint16_t status;
+	uint32_t cell_id;
+	uint16_t mode;
+	int return_code;
+	int coneval_result;
+	const char *plmn_str;
+	struct lte_lc_conn_eval_params coneval;
+	uint16_t loop;
+	uint16_t return_to_pattern;
+	uint16_t band_optimization;
+	char *pattern_buf[4];
+};
+
+static enum nrf_modem_at_scanf_variant_t nrf_modem_at_scanf_variant[5];
+static struct nrf_modem_at_scanf_return_t nrf_modem_at_scanf_return = {
+	.plmn_str = "",
+	.pattern_buf = {"", "", "", ""},
+};
+
+int nrf_modem_at_scanf(const char *cmd, const char *fmt, ...)
```

Comment:
This is a general comment. I was hoping that we could create a general mock for nrf_modem_at_scanf, and also nrf_modem_at_printf (required when variadic parameters are used) so various tests could use it.
Have you discussed this with @lemrey?

Reply:
It could be nice to have a common mock for these! Unfortunately, this would get rather large since it would need to cover all of the different use cases along with possible outputs.

Reply:
I've tried to make some draft of how the generic solution could potentially look like so that other tests could utilize it also:
https://github.com/trantanen/sdk-nrf/blob/at_scanf_cmock/tests/lib/lte_lc_api/src/lte_lc_api_test.c#L288
We would need to put this into a library, e.g., test_mock_nrf_modem_at().
I didn't have time today to complete all tests with the mock trial but haven't yet found a reason why this wouldn't work. And so far I think this would be better as it would be more generic. I can continue tomorrow to make all tests pass with this.
I don't know if @lemrey has an idea how this would be even more generic.
Obviously, the ultimate solution would be to make variadic argument support into CMock itself.

Reply:
We in portland have been talking about the difficulty in mocking our variadic nrf_modem functions too. I actually brought up creating a centralized solution on one of Noric's testing Teams channels quite a while ago, but it didn't get any response, so the Portland team decided to pursue it independently. While we've not built anything yet, we do have a ticket describing a possible generalized solution: https://projecttools.nordicsemi.no/jira/browse/IRIS-4770

Tagging @plskeggs and @tony-le-24 on this

Reply:
Generic CMock for nrf_modem_at in PR #8766 has progressed and is fairly usable already. We just need a decision on which way to go and we should make that decision before merging these PRs.